src/protocol/attribution/input.rs

```diff
@@ -0,0 +1,172 @@
+}
+
+#[derive(Debug)]
+pub struct MCAggregateCreditOutputRow<F: Field> {
+    pub breakdown_key: Vec<AdditiveShare<F>>,
+    pub credit: AdditiveShare<F>,
+}
+
+#[async_trait]
+impl<F: Field> Resharable<F> for MCAccumulateCreditInputRow<F> {
+    type Share = AdditiveShare<F>;
+
+    async fn reshare<C>(&self, ctx: C, record_id: RecordId, to_helper: Role) -> Result<Self, Error>
+    where
+        C: Context<F, Share = <Self as Resharable<F>>::Share> + Send,
+    {
+        let f_trigger_bit = ctx
+            .narrow(&AttributionResharableStep::IsTriggerReport)
+            .reshare(&self.is_trigger_report, record_id, to_helper);
+        let f_helper_bit = ctx.narrow(&AttributionResharableStep::HelperBit).reshare(
+            &self.helper_bit,
+            record_id,
+            to_helper,
+        );
+        let f_breakdown_key = self.breakdown_key.reshare(
+            ctx.narrow(&AttributionResharableStep::BreakdownKey),
+            record_id,
+            to_helper,
+        );
+        let f_value = ctx
+            .narrow(&AttributionResharableStep::TriggerValue)
+            .reshare(&self.trigger_value, record_id, to_helper);
+
+        let (breakdown_key, mut fields) = try_join(
+            f_breakdown_key,
+            try_join_all([f_trigger_bit, f_helper_bit, f_value]),
+        )
+        .await?;
+
+        Ok(MCAccumulateCreditInputRow {
+            breakdown_key,
+            is_trigger_report: fields.remove(0),
+            helper_bit: fields.remove(0),
+            trigger_value: fields.remove(0),
+        })
+    }
+}
+
+#[async_trait]
+impl<F: Field + Sized> Resharable<F> for MCCappedCreditsWithAggregationBit<F> {
+    type Share = AdditiveShare<F>;
+
+    async fn reshare<C>(&self, ctx: C, record_id: RecordId, to_helper: Role) -> Result<Self, Error>
+    where
+        C: Context<F, Share = <Self as Resharable<F>>::Share> + Send,
+    {
+        let f_helper_bit = ctx.narrow(&AttributionResharableStep::HelperBit).reshare(
+            &self.helper_bit,
+            record_id,
+            to_helper,
+        );
+        let f_aggregation_bit = ctx
+            .narrow(&AttributionResharableStep::AggregationBit)
+            .reshare(&self.aggregation_bit, record_id, to_helper);
+        let f_breakdown_key = self.breakdown_key.reshare(
+            ctx.narrow(&AttributionResharableStep::BreakdownKey),
+            record_id,
+            to_helper,
+        );
+        let f_value = ctx
+            .narrow(&AttributionResharableStep::TriggerValue)
+            .reshare(&self.credit, record_id, to_helper);
+
+        let (breakdown_key, mut fields) = try_join(
+            f_breakdown_key,
+            try_join_all([f_aggregation_bit, f_helper_bit, f_value]),
+        )
+        .await?;
+
+        Ok(MCCappedCreditsWithAggregationBit {
+            breakdown_key,
+            helper_bit: fields.remove(0),
+            aggregation_bit: fields.remove(0),
+            credit: fields.remove(0),
```

Comment:
Some day, I hope to find a better way to run these in parallel. This whole `.remove(0)` thing feels so awkward.